File: android/app/src/main/java/com/ilovetrains/app/TripDeletion.kt
package com.ilovetrains.app

const val UndoWindowMillis = 4_000L

data class PendingDeletion(val trip: SavedTrip, val index: Int, val history: List<ViewEvent>,
    val focus: FocusedJourney?, val lastAnswer: LastAnswer?, val lastTripId: String?)

fun deletionMessage(trip: SavedTrip) = "${trip.from.shortName} → ${trip.to.shortName} deleted"

fun UserData.beginDeletion(id: String): Pair<UserData, PendingDeletion>? {
    val index = trips.indexOfFirst { it.id == id }
    if (index < 0) return null
    val pending = PendingDeletion(trips[index], index, history.filter { it.tripId == id },
        focus?.takeIf { it.tripId == id }, lastAnswer?.takeIf { it.tripId == id }, lastTripId?.takeIf { it == id })
    val remaining = copy(trips = trips.filter { it.id != id }, history = history.filter { it.tripId != id },
        focus = focus?.takeIf { it.tripId != id }, lastAnswer = lastAnswer?.takeIf { it.tripId != id },
        lastTripId = lastTripId?.takeIf { it != id })
    return remaining to pending
}

fun UserData.restore(pending: PendingDeletion): UserData {
    val pair = setOf(pending.trip.from.id, pending.trip.to.id)
    if (trips.any { it.id == pending.trip.id || setOf(it.from.id, it.to.id) == pair }) return this
    val restored = trips.toMutableList().apply { add(pending.index.coerceAtMost(size), pending.trip) }
    return copy(trips = restored, history = (history + pending.history).takeLast(500),
        focus = focus ?: pending.focus, lastAnswer = lastAnswer ?: pending.lastAnswer, lastTripId = lastTripId ?: pending.lastTripId)
}

File: ios/ILoveTrains/Core/TripDeletion.swift
import Foundation

let defaultUndoWindow: Duration = .seconds(4)

struct PendingDeletion: Equatable {
    let trip: SavedTrip
    let index: Int
    let history: [ViewEvent]
    let focus: FocusedJourney?
    let lastAnswer: LastAnswer?
    let lastTripId: String?
}

func deletionMessage(_ trip: SavedTrip) -> String { "\(trip.from.shortName) → \(trip.to.shortName) deleted" }

extension UserData {
    func beginningDeletion(of id: String) -> (UserData, PendingDeletion)? {
        guard let index = trips.firstIndex(where: { $0.id == id }) else { return nil }
        let pending = PendingDeletion(
            trip: trips[index], index: index, history: history.filter { $0.tripId == id },
            focus: focus?.tripId == id ? focus : nil, lastAnswer: lastAnswer?.tripId == id ? lastAnswer : nil,
            lastTripId: lastTripId == id ? lastTripId : nil
        )
        var remaining = self
        remaining.trips.remove(at: index)
        remaining.history.removeAll { $0.tripId == id }
        if remaining.focus?.tripId == id { remaining.focus = nil }
        if remaining.lastAnswer?.tripId == id { remaining.lastAnswer = nil }
        if remaining.lastTripId == id { remaining.lastTripId = nil }
        return (remaining, pending)
    }

    func restoring(_ pending: PendingDeletion) -> UserData {
        let pair = Set([pending.trip.from.id, pending.trip.to.id])
        if trips.contains(where: { $0.id == pending.trip.id || Set([$0.from.id, $0.to.id]) == pair }) { return self }
        var restored = self
        restored.trips.insert(pending.trip, at: min(pending.index, trips.count))
        restored.history = Array((history + pending.history).suffix(500))
        restored.focus = focus ?? pending.focus
        restored.lastAnswer = lastAnswer ?? pending.lastAnswer
        restored.lastTripId = lastTripId ?? pending.lastTripId
        return restored
    }
}
